ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java

```diff
@@ -0,0 +1,381 @@
+
+    private boolean setBase(Output.Result result, Input input) {
+        boolean newEntry = false;
+        List<Output.Bucket> entries = result.getBuckets();
+        int baseEntryIndex = 0;
+        Output.Bucket baseEntry = entries.get(baseEntryIndex);
+        baseEntry.setBase(Optional.of(baseEntry));
+        for (int i = 1; i < entries.size(); i++) {
+            Output.Bucket currentEntry = entries.get(i);
+            if (input.getAnomalyStartTime().isPresent()) {
+                if (currentEntry.getEndTime() > input.getAnomalyStartTime().get()) {
+                    currentEntry.setBase(Optional.of(baseEntry));
+                    currentEntry.setCompleted(new AtomicBoolean(false));
+                    newEntry = true;
+                }
+            } else {
+                if (Math.abs(1. - currentEntry.getOverallAggValue() / baseEntry.getOverallAggValue()) > MIN_DIFF_PCT) {
+                    currentEntry.setBase(Optional.of(baseEntry));
+                    currentEntry.setCompleted(new AtomicBoolean(false));
+                    newEntry = true;
+                }
+            }
+        }
+        return newEntry;
+    }
+
+    private MultiSearchRequest newSearchRequestForOverallAggregates(Input input, AggregationBuilder agg,
+                                                                    LocalizationTimeBuckets timeBuckets) {
+        MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        timeBuckets.getAllIntervals().stream().map(i -> {
+            RangeQueryBuilder timeRangeFilter = new RangeQueryBuilder(input.getTimeFieldName())
+                    .from(i.getKey(), true)
+                    .to(i.getValue(), true);
+            BoolQueryBuilder filter = QueryBuilders.boolQuery().filter(timeRangeFilter);
+            input.getFilterQuery().ifPresent(q -> filter.filter(q));
+            SearchSourceBuilder search = new SearchSourceBuilder().size(0).query(filter).aggregation(agg);
+            SearchRequest searchRequest = new SearchRequest(new String[]{input.getIndexName()}, search);
+            return searchRequest;
+        }).forEach(multiSearchRequest::add);
+        return multiSearchRequest;
+    }
+
+    private LocalizationTimeBuckets getTimeBuckets(Input input) {
+        if ((input.getEndTime() - input.getStartTime()) < 2 * input.getMinTimeInterval()) {
+            throw new IllegalArgumentException("Time range is too short");
+        }
+
+        LocalizationTimeBuckets buckets;
+        if (input.getAnomalyStartTime().isPresent()) {
+            long anomalyStart = input.getAnomalyStartTime().get();
+            long end = Math.max(input.getEndTime(), anomalyStart + input.getMinTimeInterval());
+            int numBuckets = Math.min((int) ((end - anomalyStart) / input.getMinTimeInterval()), MAX_TIME_BUCKETS - 1);
+            long bucketInterval = (end - anomalyStart) / numBuckets;
+            long start = Math.min(input.getStartTime(), anomalyStart - bucketInterval);
```

Comment:
Should we move this line before line 335 and calculate `numBuckets` as
> Math.min((int) ((end - **start**) / input.getMinTimeInterval()), MAX_TIME_BUCKETS - 1);

Reply:
it should be anomalyStart because localization is aimed at anomalous time, which is after the given anomalyStart, thus anomaly localization. the start is used for the base bucket.